core/chains/orm.go

```diff
@@ -0,0 +1,246 @@
+	s := fmt.Sprintf(`UPDATE %s_chains SET cfg = cfg - $1 WHERE id = $2`, o.prefix)
+	res, err := o.q.Exec(s, name, chainID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to clear chain config for chain ID %v", chainID)
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected == 0 {
+		return errors.Wrapf(sql.ErrNoRows, "no chain found with ID %v", chainID)
+	}
+	return nil
+}
+
+func (o *ChainsORM[ID, CFG, CH]) DeleteChain(id ID, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+	query := fmt.Sprintf(`DELETE FROM %s_chains WHERE id = $1`, o.prefix)
+	result, err := q.Exec(query, id)
+	if err != nil {
+		return err
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected == 0 {
+		return sql.ErrNoRows
+	}
+	return nil
+}
+
+func (o *ChainsORM[ID, CFG, CH]) Chains(offset, limit int, qopts ...pg.QOpt) (chains []CH, count int, err error) {
+	err = o.q.WithOpts(qopts...).Transaction(func(q pg.Queryer) error {
+		if err = q.Get(&count, fmt.Sprintf("SELECT COUNT(*) FROM %s_chains", o.prefix)); err != nil {
+			return errors.Wrap(err, "failed to fetch chains count")
+		}
+
+		sql := fmt.Sprintf(`SELECT * FROM %s_chains ORDER BY created_at, id LIMIT $1 OFFSET $2;`, o.prefix)
+		err = q.Select(&chains, sql, pg.Limit(limit), offset)
+		return errors.Wrap(err, "failed to fetch chains")
+	}, pg.OptReadOnlyTx())
+
+	return
+}
+
+func (o *ChainsORM[ID, CFG, CH]) EnabledChains(qopts ...pg.QOpt) (chains []CH, err error) {
+	q := o.q.WithOpts(qopts...)
+	chainsSQL := fmt.Sprintf(`SELECT * FROM %s_chains WHERE enabled ORDER BY created_at, id;`, o.prefix)
+	if err = q.Select(&chains, chainsSQL); err != nil {
+		return
+	}
+	return
+}
+
+// NodesORM is a generic ORM for nodes.
+//
+// OPT: shared generic Node = db.Node[whichever.Node]
+type NodesORM[ID, NEW, N any] struct {
+	q           pg.Q
+	prefix      string
+	createNodeQ string
+}
+
+// NewNodesORM returns a NodesORM backed by q, for the table <prefix>_nodes.
+// NEW must have fields matching the column <prefix>_chain_id, as well as the given nodeCols.
+func NewNodesORM[ID, NEW, N any](q pg.Q, prefix string, nodeCols ...string) *NodesORM[ID, NEW, N] {
+	// pre-compute query for CreateNode
+	var withColon []string
+	for _, c := range nodeCols {
+		withColon = append(withColon, ":"+c)
+	}
+	query := fmt.Sprintf(`INSERT INTO %s_nodes (name, %s_chain_id, %s, created_at, updated_at)
+		VALUES (:name, :%s_chain_id, %s, now(), now())
+		RETURNING *;`, prefix, prefix, strings.Join(nodeCols, ", "), prefix, strings.Join(withColon, ", "))
+
+	return &NodesORM[ID, NEW, N]{q: q, prefix: prefix, createNodeQ: query}
+}
+
+func (o *NodesORM[ID, NEW, N]) CreateNode(data NEW, qopts ...pg.QOpt) (node N, err error) {
+	q := o.q.WithOpts(qopts...)
+	stmt, err := q.PrepareNamed(o.createNodeQ)
+	if err != nil {
+		return node, err
+	}
+	err = stmt.Get(&node, data)
```

Comment:
I'm not sure how the db packages would handle this, but it is no different than how it would have behaved in the non-generic version, no? Or do you mean that I changed something?